clients/src/main/java/org/apache/kafka/common/requests/StopReplicaRequest.java

```diff
@@ -17,149 +17,120 @@
                        Collection<TopicPartition> partitions) {
             super(ApiKeys.STOP_REPLICA, version, controllerId, controllerEpoch, brokerEpoch);
-            this.deletePartitions = deletePartitions;
-            this.partitions = partitions;
+            data = new StopReplicaRequestData()
+                .setControllerId(controllerId)
+                .setControllerEpoch(controllerEpoch)
+                .setBrokerEpoch(brokerEpoch)
+                .setDeletePartitions(deletePartitions);
+
+            if (version == 0) {
+                List<StopReplicaRequestPartitionV0> requestPartitions = partitions.stream().map(tp ->
+                    new StopReplicaRequestPartitionV0()
+                        .setTopicName(tp.topic())
+                        .setPartitionIndex(tp.partition())
+                ).collect(Collectors.toList());
+                data.setPartitionsV0(requestPartitions);
+            } else {
+                Map<String, List<Integer>> topicPartitionsMap = CollectionUtils.groupPartitionsByTopic(partitions);
+                List<StopReplicaRequestTopic> topics = topicPartitionsMap.entrySet().stream().map(entry ->
+                    new StopReplicaRequestTopic()
+                        .setName(entry.getKey())
+                        .setPartitionIndexes(entry.getValue())
+                ).collect(Collectors.toList());
+                data.setTopics(topics);
+            }
         }
 
-        @Override
         public StopReplicaRequest build(short version) {
-            return new StopReplicaRequest(controllerId, controllerEpoch, brokerEpoch,
-                    deletePartitions, partitions, version);
+            return new StopReplicaRequest(data, version);
         }
 
         @Override
         public String toString() {
-            StringBuilder bld = new StringBuilder();
-            bld.append("(type=StopReplicaRequest").
-                append(", controllerId=").append(controllerId).
-                append(", controllerEpoch=").append(controllerEpoch).
-                append(", deletePartitions=").append(deletePartitions).
-                append(", brokerEpoch=").append(brokerEpoch).
-                append(", partitions=").append(Utils.join(partitions, ",")).
-                append(")");
-            return bld.toString();
+            return data.toString();
         }
     }
 
-    private final boolean deletePartitions;
-    private final Collection<TopicPartition> partitions;
-
-    private StopReplicaRequest(int controllerId, int controllerEpoch, long brokerEpoch, boolean deletePartitions,
-                               Collection<TopicPartition> partitions, short version) {
-        super(ApiKeys.STOP_REPLICA, version, controllerId, controllerEpoch, brokerEpoch);
-        this.deletePartitions = deletePartitions;
-        this.partitions = partitions;
+    private final StopReplicaRequestData data;
+    private final Collection<TopicPartition> topicPartitions;
```

Comment:
Hmm,. can we just create `topicPartitions` as needed?  I don't think `StopReplicaRequest` needs this level of optimization, and duplicating stuff is always frustrating.